npm/networkPolicyController.go

```diff
@@ -0,0 +1,492 @@
+	netPolListerSynced cache.InformerSynced
+	workqueue          workqueue.RateLimitingInterface
+	// (TODO): networkPolController does not need to have whole NetworkPolicyManager pointer. Need to improve it
+	npMgr *NetworkPolicyManager
+	// flag to indicate default Azure NPM chain is created or not
+	isAzureNpmChainCreated bool
+}
+
+func NewNetworkPolicyController(npInformer networkinginformers.NetworkPolicyInformer, clientset kubernetes.Interface, npMgr *NetworkPolicyManager) *networkPolicyController {
+	netPolController := &networkPolicyController{
+		clientset:              clientset,
+		netPolLister:           npInformer.Lister(),
+		netPolListerSynced:     npInformer.Informer().HasSynced,
+		workqueue:              workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "NetworkPolicy"),
+		npMgr:                  npMgr,
+		isAzureNpmChainCreated: false,
+	}
+
+	npInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    netPolController.addNetworkPolicy,
+			UpdateFunc: netPolController.updateNetworkPolicy,
+			DeleteFunc: netPolController.deleteNetworkPolicy,
+		},
+	)
+	return netPolController
+}
+
+// needSync checks whether the obj is valid network policy object.
+func (c *networkPolicyController) needSync(obj interface{}) (string, error) {
+	var key string
+	_, ok := obj.(*networkingv1.NetworkPolicy)
+	if !ok {
+		return key, fmt.Errorf("cannot cast obj (%v) to network policy obj", obj)
+	}
+
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		return key, fmt.Errorf("error due to %s", err)
+	}
+
+	return key, nil
+}
+
+func (c *networkPolicyController) addNetworkPolicy(obj interface{}) {
+	key, err := c.needSync(obj)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+
+	c.workqueue.Add(key)
+}
+
+func (c *networkPolicyController) updateNetworkPolicy(old, new interface{}) {
+	key, err := c.needSync(new)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+
+	// needSync already checked validation of casting new network policy.
+	newNetPol, _ := new.(*networkingv1.NetworkPolicy)
+	oldNetPol, ok := old.(*networkingv1.NetworkPolicy)
+	if ok {
+		if oldNetPol.ResourceVersion == newNetPol.ResourceVersion {
+			// Periodic resync will send update events for all known network plicies.
+			// Two different versions of the same network policy will always have different RVs.
+			return
+		}
+	}
+
+	netPolCachedKey := util.GetNSNameWithPrefix(key)
+	c.npMgr.Lock()
+	cachedNetPolObj, netPolExists := c.npMgr.RawNpMap[netPolCachedKey]
+	c.npMgr.Unlock()
+	if netPolExists {
+		// if network policy does not have different states against lastly applied states stored in cachedNetPolObj,
+		// netPolController does not need to reconcile this update.
+		// in this updateNetworkPolicy event, newNetPol was updated with states which netPolController does not need to reconcile.
+		if isSameNetworkPolicy(cachedNetPolObj, newNetPol) {
```

Comment:
This can also be called as needSync, but I'm fine with isSameNetworkPolicy

Reply:
Now I would like to stick to isSameNetworkPolicy name since it is related to isSamePolicy function in parsePolicy.go and  ProcessedNpMap data structure. So, when revisiting this issue, I will consider suggestion.